Adding information to the LabVIEW_Debugging_Techniques.md (#9)
ni/niveristand-custom-device-handbook: docs/LabVIEW_Debugging_Techniques.md

#### LabVIEW Debugging Techniques

Custom devices are written in LabVIEW code. Therefore it’s possible to develop, test and debug the code in LabVIEW, before running the niveristand-custom-device-wizard. In other words, you can use **[LabVIEW’s built-in Debugging Techniques](https://www.ni.com/en-us/support/documentation/supplemental/12/debugging-techniques-in-labview.html)** during development, and merge the LabVIEW code into the custom device framework after it matures.

Since the custom device is one of many parts of the system definition, the behavior of the LabVIEW code within the custom device framework will likely differ from the stand-alone LabVIEW application, especially in regards to timing. As a result, you should benchmark the custom device inside of the VeriStand Engine.

Once added to the system definition, custom devices have been fully integrated into VeriStand’s context. As a result, LabVIEW’s built-in Debugging Techniques are no longer available. Several techniques are available for debugging and benchmarking the custom device.
 
#### Console Viewer

> A subcomponent of VeriStand RT Engine is the [Console Viewer](https://www.ni.com/documentation/en/veristand/latest/manual/viewing-console-output/).

Reviewer: ```suggestion
A subcomponent of VeriStand RT Engine is the **[Console Viewer](https://www.ni.com/documentation/en/veristand/latest/manual/viewing-console-output/)**.
```

### Debugging and Benchmarking

Debugging and benchmarking is a normal process of code development. There are a variety of ways to debug and benchmark custom devices.

#### LabVIEW Debugging Techniques

> Custom devices are written in LabVIEW code. Therefore it’s possible to develop, test and debug in LabVIEW before running the Custom Device Template Tool. In other words, you can use **[LabVIEW’s built-in debugging techniques](https://www.ni.com/en-us/support/documentation/supplemental/12/debugging-techniques-in-labview.html)** during development; and merge the LabVIEW code into the custom device framework after it matures.

Reviewer: ```suggestion
Custom devices are written in LabVIEW code. Therefore it’s possible to develop, test and debug the code in LabVIEW, before running the niveristand-custom-device-wizard. In other words, you can use **[LabVIEW’s built-in Debugging Techniques](https://www.ni.com/en-us/support/documentation/supplemental/12/debugging-techniques-in-labview.html)** during development; and merge the LabVIEW code into the custom device framework after it matures.
```

Since the custom device is one of many parts of the system definition, the behavior of the LabVIEW code within the custom device framework will likely differ from the stand-alone LabVIEW application, especially in regards to timing. As a result, you should benchmark the custom device inside of the VeriStand Engine.

> Once added to the system definition, custom devices have been fully integrated into VeriStand’s context. As a result, LabVIEW’s built-in debugging techniques are no longer available. Several techniques are available for debugging and benchmarking the custom device.

Reviewer: ```suggestion
Once added to the system definition, custom devices have been fully integrated into VeriStand’s context. As a result, LabVIEW’s built-in Debugging Techniques are no longer available. Several techniques are available for debugging and benchmarking the custom device.
```

 
#### Console Viewer

A subcomponent of VeriStand RT Engine is the RT Console Viewer.

> You can install it to the execution host using Measurement and Automation Explorer. When installed, the component runs a small UDP daemon allowing the operator to view the console from a utility called the RT Console Viewer. You can access the RT Console viewer from **VeriStand » Workspace » Tools » Console Viewer**.	

Reviewer: ```suggestion
You can install it to the execution host using Measurement and Automation Explorer. When installed, the component runs a small UDP daemon allowing the operator to view the console from a utility called the RT Console Viewer. You can access the RT Console viewer from **VeriStand » Tool Launcher » View Console**.	
```

> ![](images/ConsoleViewer1.jpg)

Reviewer: Please replace ConsoleViewer1.jpg with this updated picture:
![image](https://user-images.githubusercontent.com/62251693/132238422-523ba58b-6227-47ce-ac9a-7ed62795ef1f.png)

**Figure: RT Console Viewer**

The Console Viewer will show the system definition and the resulting CPU usage. The viewer is also useful for displaying debugging messages. The console viewer provides a periodic snapshot of utilization. CPU spikes and transients will probably be unobservable. If the system is very busy, it may not update the console viewer at all. You can use other debugging methods for a more accurate indication of resource utilization.

> As the name implies, the RT Console viewer is only available on real-time targets. The RT Console Viewer is also available as a stand-alone add-on to LabVIEW Real-Time. See **[NI Developer Zone Tutorial: Remotely View Console Output of Real-Time Targets](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000x4TjCAI&l=en-US)** for more information.

Reviewer: ```suggestion
As the name implies, the RT Console viewer is only available on real-time targets. The RT Console Viewer is also available as a stand-alone add-on to LabVIEW Real-Time. See **[NI Developer Zone Tutorial: Remotely View Console Output of Real-Time Targets](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000x4TjCAI&l=en-US)** for more information.

&nbsp;&nbsp;&nbsp; **Note:** You cannot use the **Console Viewer** on NI Linux Real-Time targets. Instead, connect your NI Linux Real-Time targets to a computer using a serial port to view the output.
```

#### Printing to the Console
Printing to the console is often all that’s needed to debug an application.

> #### Printing With NIVS Debug String VI

Reviewer: ```suggestion
#### Printing With "NI VeriStand - Print Debug Line" VI
```

> The recommended method of printing to the console is to use Print NIVS Debug String VI. You can download the VI from NI Community » VeriStand Add-Ons » Documents » Print VeriStand Debug String.

Reviewer: ```suggestion
The recommended method of printing to the console is to use "NI VeriStand - Print Debug Line" VI. This VI is part of Custom Device API » Utilities » Print Debug Line.vi.
```

> ![](images/NIVSdebugString.jpg)

Reviewer: Please replace the photo of the VI with this one:
![image](https://user-images.githubusercontent.com/62251693/132237918-3831c5bd-fe7e-4d5d-aa83-cf3d36b064c6.png)

> This VI works on both Windows and RT execution hosts. It has an optional input to change the color of the text. It also has an optional input to append the string to the VeriStand log file.

Reviewer: ```suggestion
This VI works on both Windows and RT execution hosts. It has an optional input (Attribute) to change the color of the text. It also has an optional input to append the string to the VeriStand log file.
```

#### Printing With ni_emb.dll

> The NIVS Debug String VI is not available in VeriStand 2009. You’ll find **ni_emb.dll** in

Reviewer: ```suggestion
```

> **<labview>\Targets\NI\RT\vi.lib**. This dll contains a stub function called **PrintStringToConsole**. Calling this function sends a string to the RT console. Configure the function to run in any thread using the C calling convention. The return type is void and it has a C String pointer input constant. You’ll find this function wrapped in a VI in the same folder in **rtutility.llb\RT Debug String.vi**. Since **ni_emb.dll** is a stub dll, it’s not necessary to deploy this VI to the RT target. The stub exists so the **PrintStringToConsole** function does not return an error when called on Windows.

Reviewer: ```suggestion
**&lt;labview&gt;\Targets\NI\RT\vi.lib**. This dll contains a stub function called **PrintStringToConsole**. Calling this function sends a string to the RT console. Configure the function to run in any thread using the C calling convention. The return type is void and it has a C String pointer input constant. You’ll find this function wrapped in a VI in the same folder in **rtutility.llb\RT Debug String.vi**. Since **ni_emb.dll** is a stub dll, it’s not necessary to deploy this VI to the RT target. The stub exists so the **PrintStringToConsole** function does not return an error when called on Windows.
```

If you do not want to call **ni_emb.dll** on a Windows OS, you can use a **[Conditional Disable Structure](https://zone.ni.com/reference/en-XX/help/371361R-01/glang/conditional_disable_structure/)** around the dll. See **[NI Tutorial: Using the Conditional Disable Structure](https://zone.ni.com/reference/en-XX/help/371361R-01/lvhowto/creating_cond_disable_struc/)** for more information.

![](images/Disable_ni_emb.jpg)	 

> Figure: Disable ni_emb.dll for non-Windows Operating Systems

Reviewer: ```suggestion

**Figure: Disable ni_emb.dll for non-Windows Operating Systems**
```

#### Distributed System Manager

You can use the NI Distributed System Manager (DSM) to monitor the CPU and memory resources of an RT target. You must install System State Publisher on the RT target. This component runs a small daemon that publishes the system state to DSM. See **[NI Distributed System Manager for LabVIEW 2018 Help](https://zone.ni.com/reference/en-XX/help/371361R-01/sysman/sysman/)** » **[System Manager Overview](https://zone.ni.com/reference/en-XX/help/372572E-01/sysman/sysman_overview/)** » **[Monitor RT target resources](https://zone.ni.com/reference/en-XX/help/372572E-01/sysman/monitoring_resources/)** for more information.

System State Publisher provides a periodic snapshot of utilization. Spikes and transients in CPU utilization will probably not be observable. If the system is very busy, it may not update DSM at all. You can use other debugging methods for a more accurate indication of resource utilization.

#### System Channels

> VeriStand includes dozens of **[system channels](https://zone.ni.com/reference/en-XX/help/372846M-01/veristand/system_channels_table/)**. System channels provide information about what’s going on under the hood of VeriStand. Several of these system channels are useful in benchmarking and debugging.

Reviewer: ```suggestion
VeriStand includes dozens of **[system channels](https://www.ni.com/documentation/en/veristand/latest/manual/system-channels/)**. System channels provide information about what’s going on under the hood of VeriStand. Several of these system channels are useful in benchmarking and debugging.
```

#### Table of Debugging and Benchmarking System Channels

```eval_rst
+-----------------+--------------------------------------------------------------------------+
|System Channel   |Description                                                               |
+=================+==========================================================================+
|HP Count         |The number of times the Primary Control Loop reported being late.         |
+-----------------+--------------------------------------------------------------------------+
|HP Loop Duration |The duration of the Primary Control Loop in nanoseconds.                  |
+-----------------+--------------------------------------------------------------------------+
|LP Count         |The number of times the Data Processing Loop reported being late.         |
+-----------------+--------------------------------------------------------------------------+
|Model Count      |The number of times the models have not completed their execution in time.|
+-----------------+--------------------------------------------------------------------------+
```
 

> If the value of the count channels increase over time, the execution host is not achieving the desired loop rates. You can use the system channels in conjunction with an **[alarm](https://zone.ni.com/reference/en-XX/help/372846M-01/veristand/add_alarms/)** or **[procedure](https://www.ni.com/documentation/en/veristand/latest/manual/add-configure-procedure/)** to handle the event.

Reviewer: ```suggestion
If the value of the count channels increase over time, the execution host is not achieving the desired loop rates. You can use the system channels in conjunction with an **[alarm](https://www.ni.com/documentation/en/veristand/latest/manual/add-configure-alarm/)** or **[procedure](https://www.ni.com/documentation/en/veristand/latest/manual/add-configure-procedure/)** to handle the event.
```

> #### System Monitor Add-on

Reviewer: ```suggestion
#### System Monitor Custom Device
```

> The VeriStand System Monitor is a Custom Device that tracks memory resources and CPU usage on an RT target running the VeriStand Engine. Set the update rate (Hz) in System Explorer to determine how often the custom device checks CPU and memory usage and sends them to the corresponding channel. The VeriStand System Monitor can only be used on an RT target. The custom device returns an error if you target it to a Windows system.

Reviewer: ```suggestion
The VeriStand [System Monitor](https://github.com/ni/niveristand-system-monitor-custom-device/releases) is a Custom Device that tracks memory resources and CPU usage on an RT target running the VeriStand Engine. Set the update rate (Hz) in System Explorer to determine how often the custom device checks CPU and memory usage and sends them to the corresponding channel. The VeriStand System Monitor can only be used on an RT target. The custom device returns an error if you target it to a Windows system.
```

> 

Reviewer: ```suggestion

#### Telemetry Custom Device

The [Telemetry Custom Device](https://github.com/ni/niveristand-telemetry-custom-device/releases) supports VeriStand benchmarking by helping you log system channels and monitor target resources. The usage data is logged to a TDMS file on the target that is running the VeriStand Engine.

```

#### Real-Time Trace Viewer

> VeriStand 2020 provides built-in support for using the **[Real-Time Trace Viewer](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtracehelp/lv_tracetoolkit_help/)** and **[Real-Time Trace Viewer VIs](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtrace/tracetoolkitvis_pal/)** to capture the timing and execution data of VI and thread events for applications running on an RT target. In LabVIEW, select **Tools»Real-Time Module»Trace Viewer** to display the Real-Time Trace Viewer. For more info on how to use Real-Time Trace Viewer, in LabVIEW go to **Help>>LabVIEW Help>>Real-Time Module>>Real-Time Trace Viewer**.

Reviewer: ```suggestion
VeriStand 2020 provides built-in support for using the **[Real-Time Trace Viewer](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtracehelp/lv_tracetoolkit_help/)** and **[Real-Time Trace Viewer VIs](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtrace/tracetoolkitvis_pal/)** to capture the timing and execution data of VI and thread events for applications running on an RT target. In a LabVIEW VI, select **Tools»Real-Time Module»Trace Viewer** to display the Real-Time Trace Viewer. For more info on how to use Real-Time Trace Viewer, in LabVIEW go to **Help>>LabVIEW Help>>Real-Time Module>>Real-Time Trace Viewer**.
```

#### Additional Debugging Options for VeriStand
Upon request, National Instrument may provide advanced debugging tools to help you resolve certain custom device issues. These tools are a last resort when all other debugging options have been exhausted. Please contact National Instruments for more information.
 

#### Table of Debugging and Benchmarking Techniques

```eval_rst

> +--------------------+-------------------+-----------+-----------------------------------------------------------------+

Reviewer: The Contents of this table should be updated with the info from here: https://www.ni.com/documentation/en/veristand/latest/manual/custom-device-benchmark-debug/

&nbsp;&nbsp;&nbsp;**Note:** You cannot use the **Console Viewer** on NI Linux Real-Time targets. Instead, connect your NI Linux Real-Time targets to a computer using a serial port to view the output.

![](images/ConsoleViewer1.jpg)

**Figure: Console Viewer**

The Console Viewer will show the system definition and the resulting CPU usage. The viewer is also useful for displaying debugging messages. The console viewer provides a periodic snapshot of utilization. CPU spikes and transients will probably be unobservable. If the system is very busy, it may not update the console viewer at all. You can use other debugging methods for a more accurate indication of resource utilization.

The Console Viewer is also available as a stand-alone add-on to LabVIEW Real-Time. See **[NI Developer Zone Tutorial: Remotely View Console Output of Real-Time Targets](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000x4TjCAI&l=en-US)** for more information.

> 

Reviewer: ```suggestion

#### Custom Error Codes

You can define custom error codes in LabVIEW and distribute the codes to VeriStand with a custom device. Copy the *custom errors.txt* file to VeriStand in the *&lt;Base&gt;\National Instruments\Shared\Errors\English* directory and add the file as a dependency in custom device and **[Custom Device XML file](https://www.ni.com/documentation/en/veristand/latest/manual/custom-device-xml/)**.

For RT targets, deploy the *errors.txt* file to the error directory on target to display error descriptions in Console Viewer.

For more information, refer to the Defining Custom Error Codes to Distribute throughout Your Application topic in the *LabVIEW Help*.

```

#### Printing With "NI VeriStand - Print Debug Line" VI

> The recommended method of printing to the console is to use "NI VeriStand - Print Debug Line" VI. This VI is part of Custom Device API » Utilities » Print Debug Line.vi.

Reviewer: ```suggestion
The recommended method of printing to the console is to use "NI VeriStand - Print Debug Line" VI. This VI is part of **Custom Device API » Utilities » Print Debug Line.vi**.
```

![](images/NIVSdebugString.jpg)

This VI works on both Windows and RT execution hosts. It has an optional input (Attribute) to change the color of the text. It also has an optional input to append the string to the VeriStand log file.

#### Printing with RT Debug String VI

The RT Debug String.VI sends a string to the standard output device. By default, this VI sends the debug string to the video port. If you have a device capable of serial redirection, this VI sends the debug string to the serial port. RT Debug String VI can be found in: **Real-Time » RT Utilities » RT Debug String.vi** LabVIEW palette.

For more information go to **[Real-Time VIs](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtvihelp/lv_real_time_vi_help/)** >> **[RT Utilities VIs](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtvihelp/rt_board_utilities_vis/)** >> **[RT Debug String VI](https://zone.ni.com/reference/en-XX/help/370715P-01/lvrtvihelp/rt_debug_strings/)** .

![](images/RT_Debug_String_VI.jpg)	 

**Figure: RT Debug String VI**

#### Telemetry Custom Device

> The [Telemetry Custom Device](https://github.com/ni/niveristand-telemetry-custom-device/releases) supports VeriStand benchmarking by helping you log system channels and monitor target resources. The usage data is logged to a TDMS file on the target that is running the VeriStand Engine.

Reviewer: ```suggestion
The **[Telemetry Custom Device](https://github.com/ni/niveristand-telemetry-custom-device/releases)** supports VeriStand benchmarking by helping you log system channels and monitor target resources. The usage data is logged to a TDMS file on the target that is running the VeriStand Engine.
```

|HP Count         |The number of times the Primary Control Loop reported being late.         |
+-----------------+--------------------------------------------------------------------------+
|HP Loop Duration |The duration of the Primary Control Loop in nanoseconds.                  |
+-----------------+--------------------------------------------------------------------------+
|LP Count         |The number of times the Data Processing Loop reported being late.         |
+-----------------+--------------------------------------------------------------------------+
|Model Count      |The number of times the models have not completed their execution in time.|
+-----------------+--------------------------------------------------------------------------+
```

If the value of the count channels increase over time, the execution host is not achieving the desired loop rates. You can use the system channels in conjunction with an **[alarm](https://www.ni.com/documentation/en/veristand/latest/manual/add-configure-alarm/)** or **[procedure](https://www.ni.com/documentation/en/veristand/latest/manual/add-configure-procedure/)** to handle the event.

#### System Monitor Custom Device

> The [System Monitor Custom Device](https://github.com/ni/niveristand-system-monitor-custom-device/releases) tracks memory resources and CPU usage on an RT target running the VeriStand Engine. Set the update rate (Hz) in System Explorer to determine how often the custom device checks CPU and memory usage and sends them to the corresponding channel. The VeriStand System Monitor can only be used on an RT target. The custom device returns an error if you target it to a Windows system.

Reviewer: ```suggestion
The **[System Monitor Custom Device](https://github.com/ni/niveristand-system-monitor-custom-device/releases)** tracks memory resources and CPU usage on an RT target running the VeriStand Engine. Set the update rate (Hz) in System Explorer to determine how often the custom device checks CPU and memory usage and sends them to the corresponding channel. The VeriStand System Monitor can only be used on an RT target. The custom device returns an error if you target it to a Windows system.
```

#### LabVIEW Debugging Techniques

Custom devices are written in LabVIEW code. Therefore it’s possible to develop, test and debug the code in LabVIEW, before running the niveristand-custom-device-wizard. In other words, you can use **[LabVIEW’s built-in Debugging Techniques](https://www.ni.com/en-us/support/documentation/supplemental/12/debugging-techniques-in-labview.html)** during development, and merge the LabVIEW code into the custom device framework after it matures.

Since the custom device is one of many parts of the system definition, the behavior of the LabVIEW code within the custom device framework will likely differ from the stand-alone LabVIEW application, especially in regards to timing. As a result, you should benchmark the custom device inside of the VeriStand Engine.
Once added to the system definition, custom devices have been fully integrated into VeriStand’s context. As a result, LabVIEW’s built-in Debugging Techniques are no longer available. Several techniques are available for debugging and benchmarking the custom device.
 
#### Console Viewer

> A subcomponent of VeriStand RT Engine is the RT Console Viewer.

Reviewer 2: no "RT" prefix

You can install it to the execution host using Measurement and Automation Explorer. When installed, the component runs a small UDP daemon allowing the operator to view the console from a utility called the RT Console Viewer. You can access the RT Console viewer from the **VeriStand Editor » Tool Launcher » View Console**.

![](images/ConsoleViewer1.jpg)

**Figure: RT Console Viewer**

The Console Viewer will show the system definition and the resulting CPU usage. The viewer is also useful for displaying debugging messages. The console viewer provides a periodic snapshot of utilization. CPU spikes and transients will probably be unobservable. If the system is very busy, it may not update the console viewer at all. You can use other debugging methods for a more accurate indication of resource utilization.

> As the name implies, the RT Console Viewer is only available on real-time targets. The RT Console Viewer is also available as a stand-alone add-on to LabVIEW Real-Time. See **[NI Developer Zone Tutorial: Remotely View Console Output of Real-Time Targets](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000x4TjCAI&l=en-US)** for more information.

Reviewer 2: It actually works on Windows.  And only will work on Windows >= VS 2021.

> &nbsp;&nbsp;&nbsp; **Note:** You cannot use the **Console Viewer** on NI Linux Real-Time targets. Instead, connect your NI Linux Real-Time targets to a computer using a serial port to view the output.

Reviewer 2: I would probably lead with this NOTE and make it more prominent since >= VS 2021 this is going to be the situation our customers will see a majority of the time (NI Linux RT targets)

#### Printing to the Console

> Printing to the console is often all that’s needed to debug an application.

Reviewer 2: This sentence is almost sarcastic; its like saying "all you need to build a house is a hammer and nails".  Technically true but practically speaking ridiculous.
I think you can remove the section + the sentence.

#### Printing With "NI VeriStand - Print Debug Line" VI
The recommended method of printing to the console is to use "NI VeriStand - Print Debug Line" VI. This VI is part of Custom Device API » Utilities » Print Debug Line.vi.

![](images/NIVSdebugString.jpg)

This VI works on both Windows and RT execution hosts. It has an optional input (Attribute) to change the color of the text. It also has an optional input to append the string to the VeriStand log file.

#### Printing With ni_emb.dll

> You’ll find **ni_emb.dll** in **&lt;labview&gt;\Targets\NI\RT\vi.lib**. This dll contains a stub function called **PrintStringToConsole**. Calling this function sends a string to the RT console. Configure the function to run in any thread using the C calling convention. The return type is void and it has a C String pointer input constant. You’ll find this function wrapped in a VI in the same folder in **rtutility.llb\RT Debug String.vi**. Since **ni_emb.dll** is a stub dll, it’s not necessary to deploy this VI to the RT target. The stub exists so the **PrintStringToConsole** function does not return an error when called on Windows.

Reviewer 2: we should not be talking about call library nodes that call .dlls.  those won't work on linux, and it looks like there's no ni_emb equivalent on linux anyway.  it is an outdated implementation detail that has been buried under another reentrant VI anyway (NI_SystemLogging.lvlib:Write to System Log.vi)

we should instead advise calling the RT Debug String.vi  and update its location to be correct, which is at: 
C:\Program Files (x86)\National Instruments\LabVIEW 2020\Targets\NI\RT\vi.lib\rtutility.llb\RT Debug String.vi

this VI will work on all RT platforms and means we don't have to talk about conditional disable structures at all.

If you do not want to call **ni_emb.dll** on a Windows OS, you can use a **[Conditional Disable Structure](https://zone.ni.com/reference/en-XX/help/371361R-01/glang/conditional_disable_structure/)** around the dll. See **[NI Tutorial: Using the Conditional Disable Structure](https://zone.ni.com/reference/en-XX/help/371361R-01/lvhowto/creating_cond_disable_struc/)** for more information.

![](images/Disable_ni_emb.jpg)	 

**Figure: Disable ni_emb.dll for non-Windows Operating Systems**

> 

Reviewer 2: I would add a new section here for using the Telemetery Custom Device before mentioning the DSM tool.

#### Telemetry Custom Device

The [Telemetry Custom Device](https://github.com/ni/niveristand-telemetry-custom-device/releases) supports VeriStand benchmarking by helping you log system channels and monitor target resources. The usage data is logged to a TDMS file on the target that is running the VeriStand Engine.

#### Real-Time Trace Viewer

VeriStand 2020 provides built-in support for using the **[Real-Time Trace Viewer](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtracehelp/lv_tracetoolkit_help/)** and **[Real-Time Trace Viewer VIs](https://zone.ni.com/reference/en-XX/help/370715P-01/lvtrace/tracetoolkitvis_pal/)** to capture the timing and execution data of VI and thread events for applications running on an RT target. In a LabVIEW VI, select **Tools»Real-Time Module»Trace Viewer** to display the Real-Time Trace Viewer. For more info on how to use Real-Time Trace Viewer, in LabVIEW go to **Help>>LabVIEW Help>>Real-Time Module>>Real-Time Trace Viewer**.

> 

Reviewer 2: In the doc help, there's also a listing for Custom Error Codes.  It is described there, but maybe to be consistent there, this document should have a section that links out to it?

#### Additional Debugging Options for VeriStand
Upon request, National Instrument may provide advanced debugging tools to help you resolve certain custom device issues. These tools are a last resort when all other debugging options have been exhausted. Please contact National Instruments for more information.
 

#### Table of Debugging and Benchmarking Techniques

You can use tools provided by LabVIEW and VeriStand to benchmark and debug your custom device.

> You can use the following table that displays various tools you can use to **[benchmark and debug VeriStand custom devices](https://www.ni.com/documentation/en/veristand/20.5/manual/custom-device-benchmark-debug/)**.

Reviewer 2: this shouldn't link to the 20.5 version of the docs, but to latest intead.
